linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -424,8 +424,12 @@ final class Emitter(config: Emitter.Config, prePrinter: Emitter.PrePrinter) {
       x._1
     }
 
-    val classTreeCache =
-      extractChanged(classCache.getCache(linkedClass.version))
+    val classTreeCache = extractChanged {
+      classCache.getCache(linkedClass.version,
+          hasInstances = linkedClass.hasInstances,
+          hasRuntimeTypeInfo = linkedClass.hasRuntimeTypeInfo,
+          hasInstanceTests = linkedClass.hasInstanceTests)
```

Comment:
The fact that the usage of this is hidden behind `needInstanceTests` bothers me quite a bit.

Have you considered passing the necessary linking info individually?
IIUC this would get us almost to the state where `linkedClass` as a whole doesn't escape `genClass` anymore. (`extractInlineableInit` and `needStaticInitialization` would need to be adjusted, but it seems to be trivial).

That in turn would allow us to annotate all fields we pass from `linkedClass` with where they are invalidated, hopefully reducing the overall likelihood of errors.

I might spend some time annotating all usages to avoid problems like this.